### ISSUE TITLE
Make the poly domains sympifiable

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#18672]
I think there is consensus that it would be good if the poly domains could be sympified and had a corresponding Basic representation. This would make it possible for Poly to have all Basic args with the domain included. Currently the args of Poly don't include the domain e.g.:
```julia
In [1]: Poly(x, x, domain='QQ')                                                                                                                
Out[1]: Poly(x, x, domain='QQ')

In [2]: Poly(x, x, domain='QQ').args                                                                                                           
Out[2]: (x, x)
```
This means that Poly instances do not properly satisfy the func-args invariant since
```julia
In [4]: p                                                                                                                                      
Out[4]: Poly(x, x, domain='QQ')

In [5]: p.func(*p.args)                                                                                                                        
Out[5]: Poly(x, x, domain='ZZ')

In [6]: p.func(*p.args) == p                                                                                                                   
Out[6]: False
```
Since the domain is not Basic it cannot be directly part of the args though.

One solution is to have Basic objects that correspond to the various non-basic domain objects and to make it so that sympify understands how to convert them. This would look something like
```julia
In [7]: sympify(ZZ) 
BasicIntegerRing
```

It's not clear to me though what the corresponding class hierarchy should be or what different domains should be exposed at the Basic level.

For example some of the domains are selected differently depending on `GROUND_TYPES` (whether gmpy is installed) so there is e.g. GMPYIntegerRing and PythonIntegerRing. Is that just an implementation detail? Should both of those correspond to the same Basic object?

It seems to me that what is needed is some singletons like
```
BasicIntegerRing
BasicRationalField
BasicRealField
BasicComplexField
```
Then there need to be non-Atom types like
```
BasicPolynomialRing
BasicAlgebraicField
```
After that I'm not sure what else would be needed...

**Comment by oscarbenjamin** [comment]
Also the Domain classes have a lot of methods. Which of those would be relevant to a BasicDomain class?

**Comment by sylee957** [comment]
Why should everything be prefixed with `Basic`. Is it because all of the non-basic types that were served before in the previous versions?

**Comment by oscarbenjamin** [comment]
I don't have strong opinions about the names but yes I prefixed them with Basic because those names are already used for the non-Basic domain types which are also exposed at top level e.g. `sympy.PolynomialRing`.

**Comment by jksuom** [comment]
I have been thinking of a more light-weight solution. `Poly` could have a third positional parameter `domsym=None`. If it is not `None`, it would be a symbol with name like 'ZZ', and then `Poly.__new__` would add the keyword `domain=domsym.name`.

**Comment by oscarbenjamin** [comment]
Is there any value in being able to substitute in the domain like
```julia
In [12]: Poly(x, y)
Out[12]: Poly(x, y, domain='ZZ[x]')

In [13]: Poly(x, y).subs(y, z)
Out[13]: Poly(x, z, domain='ZZ[x]')

In [14]: Poly(x, y).subs(x, z)
Out[14]: z
```
So e.g. the last example could be `Poly(z, y, domain='ZZ[z]')`. That would work if there was some object for `ZZ[x]` that holds `x` as an arg.

Can all domains be represented as strings and are such representations unique? For example we have
```julia
In [35]: y1 = Symbol('y', positive=True)

In [36]: y2 = Symbol('y', real=True)

In [37]: Poly(y1*x, x).domain == Poly(y2*x, x).domain
Out[37]: False

In [38]: str(Poly(y1*x, x).domain) == str(Poly(y2*x, x).domain)
Out[38]: True
```

**Comment by jksuom** [comment]
> Can all domains be represented as strings and are such representations unique

Probably not for the moment but I think that it would be possible to develop suitable representations.
Note that the variables of the so-called composite domains, e.g. `y1` and `y2` above, are not expected to have assumptions. They should be regarded as pure transcendentals.

**Comment by oscarbenjamin** [comment]
> They should be regarded as pure transcendentals.

Yes, but the point is that distinct symbols can have the same string name. We can have the same with `Dummy` e.g.:
```julia
In [55]: y1 = Dummy('y')

In [56]: y2 = Dummy('y')

In [57]: Poly(y1*x, x) * Poly(y2*x, x)
Out[57]: Poly(_y*_y*x**2, x, domain='ZZ[_y,_y]')
```

**Comment by jksuom** [comment]
> distinct symbols can have the same string name.

Yes, but I am not sure that such choices would be unavoidable in a single context.
